verbit/streaming_client.py

```diff
@@ -342,3 +448,109 @@ def _get_ws_connect_query_string(media_config: MediaConfig, response_types: Resp
+              1. WebSocketException: from remote WebSocket connection
+              2. ConnectionError: when WebSocket is on the same local machine, seems to be using Pipes etc
+              3. TimeoutError: From OS-level 'select()' or similar operations, for example when physically disconnecting for a network results in a timeout
+              4. socket.timeout/socket.gaierror: merged into '3' in python 3.10, required in earlier versions
+
+        """
+
+        ended = False
+
+        while not ended:
+
+            try:
+                yield from response_generator
+
+                # response_generator exhausted without any exceptions
+                ended = True
+
+            except (ConnectionError, WebSocketException, TimeoutError, socket.timeout, socket.gaierror) as connection_error:
+                self._logger.debug(f'caught {type(connection_error).__name__}: {connection_error}, will try reconnect-and-continue')
+
+                # wait for other threads, that still access the web-socket to fail and stop
+                self._wait_for_thread(self._media_sender_thread, timeout_step=0.1, global_timeout=1.0, logger=self._logger)
+
+                # try reconnecting as soon as the exception's been caught
+                # and keep on yielding from the same generator
+
+                self._logger.debug('Will try reconnecting')
+                try:
+                    response_generator = super().start_stream(self._media_generator, self._media_config, self._response_types)
+                except tenacity.RetryError:
+                    self._logger.error('Gave up retrying during reconnection. stopping _reconnect_generator()')
+                    ended = True
+                    # do not raise further
+
+                except Exception:
+                    self._logger.exception('caught during attempting to reconnect, stopping _reconnect_generator()')
+                    ended = True
+                    raise
+
+            # any other exception ends iteration
+            except Exception as ex:
+                self._logger.debug(f'Stack trace from reconnect_generator():, logging and ignoring exception {type(ex)}:{ex} .', exc_info=True)
+                # stop generator
+                ended = True
+
+
+    @staticmethod
+    def _wait_for_thread(thread: Thread , timeout_step, global_timeout=10000.0, logger=None):
+        """Join thread with logging on result"""
+        waiting_for = 0.0
+        while True:
+            thread.join(timeout=timeout_step)
+            if thread.is_alive():
+                waiting_for += timeout_step
+                if logger:
+                    logger.debug(f'{thread.name} not closing, {waiting_for=} seconds...')
+                if waiting_for >= global_timeout:
+                    return False
+            else:
+                if logger:
+                    logger.debug(f'{thread.name} closed, {waiting_for=} seconds...')
+                return True
+
```

Comment:
^ might not be really needed, or moves to some 'utilities' files

Reply:
No need to create a 'utils' module for now. I refactored this function a bit. Please have a look.